File: src/test/java/com/davidbarsky/ExperimentLoggerTest.java
package com.davidbarsky;

import com.davidbarsky.schedulers.EdgeZero;
import org.junit.Test;

import java.io.File;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.List;

import static org.junit.Assert.*;

public class ExperimentLoggerTest {

    @Test
    public void toCSV() throws Exception {
        List<ExperimentResult> experimentResults = ExperimentRunner.runSeries(new EdgeZero(), 10, 12);
        String csv = ExperimentLogger.toCSV(experimentResults);

        // There's probably a better way to test formatting. Cost is non-deterministic
        System.out.println(csv);
        assertNotNull(csv);
    }

    @Test
    public void writeToFile() throws Exception {
        List<ExperimentResult> experimentResults = ExperimentRunner.runSeries(new EdgeZero(), 10, 12);
        String testCSV = ExperimentLogger.toCSV(experimentResults);

        File file = File.createTempFile("output", "txt");
        file.deleteOnExit();

        ExperimentLogger.writeToFile(experimentResults, file);
        List<String> strings = Files.readAllLines(file.toPath());
        String testActual = String.join("\n", strings);

        assertEquals(testCSV, testActual);
    }
}

File: src/test/java/quick/sirens/experiments/ExperimentLoggerTest.scala
package quick.sirens.experiments

import java.io.File
import java.nio.file.Files

import org.junit.Assert._
import org.junit.Test
import sirens.experiments.{ExperimentLogger, ExperimentResult, ExperimentRunner, GraphGenerator}
import sirens.models.states.MachineType
import sirens.models.data.{LatencyBounds, NetworkingBounds}
import sirens.schedulers.EdgeZero

class ExperimentLoggerTest {
  @Test
  @throws[Exception]
  def toCSV(): Unit = {
    val graph = GraphGenerator.genericGraph(30)
    val experimentResult: ExperimentResult =
      ExperimentRunner.runExperiment(new EdgeZero,
                                     graph.size,
                                     graph,
                                     MachineType.SMALL,
                                     NetworkingBounds(10, 60),
                                     LatencyBounds(10, 60))
    val csv: String = ExperimentLogger.toCSV(experimentResult :: Nil)
    // We're checking the first parts of the CSV fit the pattern of "EdgeZero,SMALL". Numbers that follow
    // are non-deterministic
    assertEquals("EdgeZero,SMALL", csv.substring(0, 14))
  }

  @Test
  @throws[Exception]
  def writeToFile(): Unit = {
    val graph = GraphGenerator.genericGraph(30)
    val experimentResult: ExperimentResult =
      ExperimentRunner.runExperiment(new EdgeZero,
                                     graph.size,
                                     graph,
                                     MachineType.SMALL,
                                     NetworkingBounds(10, 60),
                                     LatencyBounds(10, 60))
    val testCSV: String = ExperimentLogger.toCSV(experimentResult :: Nil)

    val file: File = File.createTempFile("output", "txt")
    file.deleteOnExit()
    ExperimentLogger.writeToFile(experimentResult :: Nil, file)
    val testActual: String = String.join("\n", Files.readAllLines(file.toPath))
    assertEquals(testCSV, testActual)
  }
}
